# H2O GBM Tuning Tutorial for Python

### Arno Candel, PhD, Chief Architect, H2O.ai
### Ported to Python by Navdeep Gill, M.S., Hacker/Data Scientist, H2O.ai

In this tutorial, we show how to build a well-tuned H2O GBM model for a supervised classification task. We specifically don't focus on feature engineering and use a small dataset to allow you to reproduce these results in a few minutes on a laptop. This script can be directly transferred to datasets that are hundreds of GBs large and H2O clusters with dozens of compute nodes.

You can download the source [from H2O's github repository](https://github.com/h2oai/h2o-3/blob/master/h2o-docs/src/product/tutorials/gbm/gbmTuning.ipynb).

Ports to [R Markdown](https://github.com/h2oai/h2o-3/blob/master/h2o-docs/src/product/tutorials/gbm/gbmTuning.Rmd) and [Flow UI](https://raw.githubusercontent.com/h2oai/h2o-3/master/h2o-docs/src/product/flow/packs/examples/GBM_TuningGuide.flow) (now part of Examples Flows) are available as well.

## Installation of the H2O Python Package
Either download H2O from [H2O.ai's website](http://h2o.ai/download) or install the latest version of H2O into Python with the following set of commands:

Install dependencies from command line (prepending with `sudo` if needed):

```
[sudo] pip install -U requests
[sudo] pip install -U tabulate
[sudo] pip install -U future
[sudo] pip install -U six
```

The following command removes the H2O module for Python.
```
[sudo] pip uninstall h2o
```

Next, use pip to install this version of the H2O Python module.
```
[sudo] pip install http://h2o-release.s3.amazonaws.com/h2o/rel-turchin/8/Python/h2o-3.8.2.8-py2.py3-none-any.whl
```

## Launch an H2O cluster on localhost

In [1]:
import h2o
import numpy as np
import math
h2o.init(nthreads=-1, strict_version_check=True)
## optional: connect to a running H2O cluster
#h2o.init(ip="mycluster", port=55555)



No instance found at ip and port: localhost:54321. Trying to start local jar...


JVM stdout: /var/folders/fm/9q0qk_xs0wd07zr2v6d5mfv40000gn/T/tmpm9dEgv/h2o_arno_started_from_python.out
JVM stderr: /var/folders/fm/9q0qk_xs0wd07zr2v6d5mfv40000gn/T/tmp_ublea/h2o_arno_started_from_python.err
Using ice_root: /var/folders/fm/9q0qk_xs0wd07zr2v6d5mfv40000gn/T/tmpMuv0aE


Java Version: java version "1.8.0_60"
Java(TM) SE Runtime Environment (build 1.8.0_60-b27)
Java HotSpot(TM) 64-Bit Server VM (build 25.60-b23, mixed mode)


Starting H2O JVM and connecting: .............. Connection successful!


/usr/local/lib/python2.7/site-packages/IPython/core/formatters.py:92: DeprecationWarning: DisplayFormatter._ipython_display_formatter_default is deprecated: use @default decorator instead.
  def _ipython_display_formatter_default(self):
/usr/local/lib/python2.7/site-packages/IPython/core/formatters.py:98: DeprecationWarning: DisplayFormatter._formatters_default is deprecated: use @default decorator instead.
  def _formatters_default(self):
/usr/local/lib/python2.7/site-packages/IPython/core/formatters.py:677: DeprecationWarning: PlainTextFormatter._deferred_printers_default is deprecated: use @default decorator instead.
  def _deferred_printers_default(self):
/usr/local/lib/python2.7/site-packages/IPython/core/formatters.py:669: DeprecationWarning: PlainTextFormatter._singleton_printers_default is deprecated: use @default decorator instead.
  def _singleton_printers_default(self):
/usr/local/lib/python2.7/site-packages/IPython/core/formatters.py:672: DeprecationWarning: PlainTextFormat

------------------------------  -----------------------------------
H2O cluster uptime:             1 seconds 514 milliseconds
H2O cluster version:            3.8.3.3
H2O cluster name:               H2O_started_from_python_arno_hfo986
H2O cluster total nodes:        1
H2O cluster total free memory:  3.56 GB
H2O cluster total cores:        8
H2O cluster allowed cores:      8
H2O cluster healthy:            True
H2O Connection ip:              127.0.0.1
H2O Connection port:            54321
H2O Connection proxy:
Python Version:                 2.7.11
------------------------------  -----------------------------------

## Import the data into H2O 
Everything is scalable and distributed from now on. All processing is done on the fully multi-threaded and distributed H2O Java-based backend and can be scaled to large datasets on large compute clusters.
Here, we use a small public dataset ([Titanic](https://stat.ethz.ch/R-manual/R-devel/library/datasets/html/Titanic.html)), but you can use datasets that are hundreds of GBs large.

In [2]:
## 'path' can point to a local file, hdfs, s3, nfs, Hive, directories, etc.
df = h2o.import_file(path = "http://s3.amazonaws.com/h2o-public-test-data/smalldata/gbm_test/titanic.csv")
print df.dim
print df.head
print df.tail
print df.describe

## pick a response for the supervised problem
response = "survived"

## the response variable is an integer, we will turn it into a categorical/factor for binary classification
df[response] = df[response].asfactor()           

## use all other columns (except for the name & the response column ("survived")) as predictors
predictors = df.columns
del predictors[1:3]
print predictors


Parse Progress: [##################################################] 100%
[1309, 14]


pclass survived name sex age sibsp parch ticket fare cabin embarked boat body home.dest 
 1 1 Allen Miss. Elisabeth Walton female 29 0 0 24160 211.338 B5 S 2 nan St Louis MO 
 1 1 Allison Master. Hudson Trevor male 0.9167 1 2 113781 151.55 C22 C26 S 11 nan Montreal PQ / Chesterville ON 
 1 0 Allison Miss. Helen Loraine female 2 1 2 113781 151.55 C22 C26 S nan nan Montreal PQ / Chesterville ON 
 1 0 Allison Mr. Hudson Joshua Creighton male 30 1 2 113781 151.55 C22 C26 S nan 135 Montreal PQ / Chesterville ON 
 1 0 Allison Mrs. Hudson J C (Bessie Waldo Daniels) female 25 1 2 113781 151.55 C22 C26 S nan nan Montreal PQ / Chesterville ON 
 1 1 Anderson Mr. Harry male 48 0 0 19952 26.55 E12 S 3 nan New York NY 
 1 1 Andrews Miss. Kornelia Theodosia female 63 1 0 13502 77.9583 D7 S 10 nan Hudson NY 
 1 0 Andrews Mr. Thomas Jr male 39 0 0 112050 0 A36 S nan nan Belfast NI 
 1 1 Appleton Mrs. Edward Dale (Charlotte Lamson) female 53 2 0 11769 51.4792 C101 S nan nan Bayside Queens NY 
 1 0 Artagaveytia Mr. Ramon male 71 0 0 nan 49.5042 C nan 22 Montevideo Uruguay

<bound method H2OFrame.head of >


pclass survived name sex age sibsp parch ticket fare cabin embarked boat body home.dest 
 1 1 Allen Miss. Elisabeth Walton female 29 0 0 24160 211.338 B5 S 2 nan St Louis MO 
 1 1 Allison Master. Hudson Trevor male 0.9167 1 2 113781 151.55 C22 C26 S 11 nan Montreal PQ / Chesterville ON 
 1 0 Allison Miss. Helen Loraine female 2 1 2 113781 151.55 C22 C26 S nan nan Montreal PQ / Chesterville ON 
 1 0 Allison Mr. Hudson Joshua Creighton male 30 1 2 113781 151.55 C22 C26 S nan 135 Montreal PQ / Chesterville ON 
 1 0 Allison Mrs. Hudson J C (Bessie Waldo Daniels) female 25 1 2 113781 151.55 C22 C26 S nan nan Montreal PQ / Chesterville ON 
 1 1 Anderson Mr. Harry male 48 0 0 19952 26.55 E12 S 3 nan New York NY 
 1 1 Andrews Miss. Kornelia Theodosia female 63 1 0 13502 77.9583 D7 S 10 nan Hudson NY 
 1 0 Andrews Mr. Thomas Jr male 39 0 0 112050 0 A36 S nan nan Belfast NI 
 1 1 Appleton Mrs. Edward Dale (Charlotte Lamson) female 53 2 0 11769 51.4792 C101 S nan nan Bayside Queens NY 
 1 0 Artagaveytia Mr. Ramon male 71 0 0 nan 49.5042 C nan 22 Montevideo Uruguay

<bound method H2OFrame.tail of >


pclass survived name sex age sibsp parch ticket fare cabin embarked boat body home.dest 
 1 1 Allen Miss. Elisabeth Walton female 29 0 0 24160 211.338 B5 S 2 nan St Louis MO 
 1 1 Allison Master. Hudson Trevor male 0.9167 1 2 113781 151.55 C22 C26 S 11 nan Montreal PQ / Chesterville ON 
 1 0 Allison Miss. Helen Loraine female 2 1 2 113781 151.55 C22 C26 S nan nan Montreal PQ / Chesterville ON 
 1 0 Allison Mr. Hudson Joshua Creighton male 30 1 2 113781 151.55 C22 C26 S nan 135 Montreal PQ / Chesterville ON 
 1 0 Allison Mrs. Hudson J C (Bessie Waldo Daniels) female 25 1 2 113781 151.55 C22 C26 S nan nan Montreal PQ / Chesterville ON 
 1 1 Anderson Mr. Harry male 48 0 0 19952 26.55 E12 S 3 nan New York NY 
 1 1 Andrews Miss. Kornelia Theodosia female 63 1 0 13502 77.9583 D7 S 10 nan Hudson NY 
 1 0 Andrews Mr. Thomas Jr male 39 0 0 112050 0 A36 S nan nan Belfast NI 
 1 1 Appleton Mrs. Edward Dale (Charlotte Lamson) female 53 2 0 11769 51.4792 C101 S nan nan Bayside Queens NY 
 1 0 Artagaveytia Mr. Ramon male 71 0 0 nan 49.5042 C nan 22 Montevideo Uruguay

<bound method H2OFrame.describe of >
[u'pclass', u'sex', u'age', u'sibsp', u'parch', u'ticket', u'fare', u'cabin', u'embarked', u'boat', u'body', u'home.dest']


From now on, everything is generic and directly applies to most datasets. We assume that all feature engineering is done at this stage and focus on model tuning. For multi-class problems, you can use `h2o.logloss()` or `h2o.confusion_matrix()` instead of `h2o.auc()` and for regression problems, you can use `h2o.mean_residual_deviance()` or `h2o.mse()`.

## Split the data for Machine Learning
We split the data into three pieces: 60% for training, 20% for validation, 20% for final testing. 
Here, we use random splitting, but this assumes i.i.d. data. If this is not the case (e.g., when events span across multiple rows or data has a time structure), you'll have to sample your data non-randomly.

In [4]:
train, valid, test = df.split_frame(
    ratios=[0.6,0.2], 
    seed=1234, 
    destination_frames=['train.hex','valid.hex','test.hex']
)

## Establish baseline performance
As the first step, we'll build some default models to see what accuracy we can expect. Let's use the [AUC metric](http://mlwiki.org/index.php/ROC_Analysis) for this demo, but you can use `h2o.logloss()` and `stopping_metric="logloss"` as well. It ranges from 0.5 for random models to 1 for perfect models.


The first model is a default GBM, trained on the 60% training split

In [5]:
#We only provide the required parameters, everything else is default
gbm = h2o.H2OGradientBoostingEstimator()
gbm.train(x=predictors, y=response, training_frame=train)

## Show a detailed model summary
print gbm


gbm Model Build Progress: [##################################################] 100%
Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Method
Model Key:  GBM_model_python_1468807556132_1

Model Summary: 


number_of_trees    number_of_internal_trees    model_size_in_bytes    min_depth    max_depth    mean_depth    min_leaves    max_leaves    mean_leaves
--  -----------------  --------------------------  ---------------------  -----------  -----------  ------------  ------------  ------------  -------------
    50                 50                          17181                  2            5            4.92          3             21            9.14



ModelMetricsBinomial: gbm
** Reported on train data. **

MSE: 0.0406797359799
R^2: 0.828341496541
LogLoss: 0.168135896237
Mean Per-Class Error: 0.0438517398512
AUC: 0.988222972832
Gini: 0.976445945665

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.40045091311: 


0    1    Error    Rate
-----  ---  ---  -------  ------------
0      472  7    0.0146   (7.0/479.0)
1      22   279  0.0731   (22.0/301.0)
Total  494  286  0.0372   (29.0/780.0)


Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.400451     0.950596  192
max f2                       0.199583     0.945596  205
max f0point5                 0.457101     0.968772  189
max accuracy                 0.433971     0.962821  191
max precision                0.979959     1         0
max recall                   0.0829058    1         251
max specificity              0.979959     1         0
max absolute_MCC             0.433971     0.92172   191
max min_per_class_accuracy   0.209993     0.939457  203
max mean_per_class_accuracy  0.400451     0.956148  192


Gains/Lift Table: Avg response rate: 38.59 %



group    cumulative_data_fraction    lower_threshold    lift       cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  ---------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0102564                   0.963035           2.59136    2.59136            1                1                           0.0265781       0.0265781                  159.136   159.136
    2        0.0217949                   0.956877           2.59136    2.59136            1                1                           0.0299003       0.0564784                  159.136   159.136
    3        0.0307692                   0.954319           2.59136    2.59136            1                1                           0.0232558       0.0797342                  159.136   159.136
    4        0.0410256                   0.951903           2.59136    2.59136            1                1                           0.0265781       0.106312                   159.136   159.136
    5        0.05                        0.951191           2.59136    2.59136            1                1                           0.0232558       0.129568                   159.136   159.136
    6        0.107692                    0.947791           2.59136    2.59136            1                1                           0.149502        0.27907                    159.136   159.136
    7        0.15                        0.946005           2.59136    2.59136            1                1                           0.109635        0.388704                   159.136   159.136
    8        0.2                         0.943196           2.59136    2.59136            1                1                           0.129568        0.518272                   159.136   159.136
    9        0.301282                    0.837355           2.59136    2.59136            1                1                           0.262458        0.780731                   159.136   159.136
    10       0.4                         0.204109           1.61539    2.3505             0.623377         0.907051                    0.159468        0.940199                   61.5395   135.05
    11       0.501282                    0.128048           0.393625   1.95512            0.151899         0.754476                    0.0398671       0.980066                   -60.6375  95.512
    12       0.601282                    0.0844031          0.0996678  1.64654            0.0384615        0.635394                    0.00996678      0.990033                   -90.0332  64.6537
    13       0.7                         0.0779819          0.100962   1.42857            0.038961         0.551282                    0.00996678      1                          -89.9038  42.8571
    14       0.8                         0.0647129          0          1.25               0                0.482372                    0               1                          -100      25
    15       0.978205                    0.0520278          0          1.02228            0                0.394495                    0               1                          -100      2.22805
    16       1                           0.0490107          0          1                  0                0.385897                    0               1                          -100      0



Scoring History: 


timestamp            duration    number_of_trees    training_MSE     training_logloss    training_AUC    training_lift    training_classification_error
---  -------------------  ----------  -----------------  ---------------  ------------------  --------------  ---------------  -------------------------------
     2016-07-17 19:06:34  0.020 sec   0.0                0.236980604865   0.66687754039       0.5             1.0              0.614102564103
     2016-07-17 19:06:34  0.133 sec   1.0                0.206485545277   0.603360501805      0.88511156271   2.59136212625    0.0897435897436
     2016-07-17 19:06:34  0.171 sec   2.0                0.182138733238   0.553097780707      0.885142773913  2.59136212625    0.0884615384615
     2016-07-17 19:06:34  0.202 sec   3.0                0.162571389022   0.512261184327      0.885142773913  2.59136212625    0.0884615384615
     2016-07-17 19:06:34  0.228 sec   4.0                0.146780719947   0.478503019475      0.885142773913  2.59136212625    0.0884615384615
---  ---                  ---         ---                ---              ---                 ---             ---              ---
     2016-07-17 19:06:35  1.030 sec   46.0               0.0406763088862  0.168137585046      0.988222972832  2.59136212625    0.0371794871795
     2016-07-17 19:06:35  1.049 sec   47.0               0.0406772863879  0.168137022774      0.988222972832  2.59136212625    0.0371794871795
     2016-07-17 19:06:35  1.068 sec   48.0               0.0406781794184  0.168136566792      0.988222972832  2.59136212625    0.0371794871795
     2016-07-17 19:06:35  1.085 sec   49.0               0.0406789941191  0.168136197404      0.988222972832  2.59136212625    0.0371794871795
     2016-07-17 19:06:35  1.103 sec   50.0               0.0406797359799  0.168135896237      0.988222972832  2.59136212625    0.0371794871795


See the whole table with table.as_data_frame()

Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
boat        628.094                1                    0.642285
home.dest   147.919                0.235505             0.151261
cabin       129.448                0.206096             0.132372
sex         50.248                 0.0800007            0.0513833
ticket      11.9593                0.0190406            0.0122295
age         5.88323                0.00936679           0.00601615
body        1.74199                0.00277345           0.00178134
embarked    1.06386                0.00169379           0.0010879
parch       0.886553               0.0014115            0.000906583
fare        0.530628               0.000844822          0.000542617
sibsp       0.131104               0.000208732          0.000134066
pclass      0                      0                    0

In [6]:
## Get the AUC on the validation set
perf = gbm.model_performance(valid)
print perf.auc()

0.948013524937


The AUC is over 94%, so this model is highly predictive!

The second model is another default GBM, but trained on 80% of the data (here, we combine the training and validation splits to get more training data), and cross-validated using 4 folds.
Note that cross-validation takes longer and is not usually done for really large datasets.

In [8]:
## rbind() makes a copy here, so it's better to use split_frame with `ratios = c(0.8)` instead above
cv_gbm = h2o.H2OGradientBoostingEstimator(nfolds = 4, seed = 0xDECAF)
cv_gbm.train(x = predictors, y = response, training_frame = train.rbind(valid))


gbm Model Build Progress: [##################################################] 100%


We see that the cross-validated performance is similar to the validation set performance:

In [9]:
## Show a detailed summary of the cross validation metrics
## This gives you an idea of the variance between the folds
cv_summary = cv_gbm.cross_validation_metrics_summary().as_data_frame()
#print(cv_summary) ## Full summary of all metrics
#print(cv_summary.iloc[4]) ## get the row with just the AUCs

## Get the cross-validated AUC by scoring the combined holdout predictions.
## (Instead of taking the average of the metrics across the folds)
perf_cv = cv_gbm.model_performance(xval=True)
print perf_cv.auc()

0.948595146871


Next, we train a GBM with "I feel lucky" parameters.
We'll use early stopping to automatically tune the number of trees using the validation AUC. 
We'll use a lower learning rate (lower is always better, just takes more trees to converge).
We'll also use stochastic sampling of rows and columns to (hopefully) improve generalization.

In [10]:
gbm_lucky = h2o.H2OGradientBoostingEstimator(
  ## more trees is better if the learning rate is small enough 
  ## here, use "more than enough" trees - we have early stopping
  ntrees = 10000,                                                            

  ## smaller learning rate is better (this is a good value for most datasets, but see below for annealing)
  learn_rate = 0.01,                                                         

  ## early stopping once the validation AUC doesn't improve by at least 0.01% for 5 consecutive scoring events
  stopping_rounds = 5, stopping_tolerance = 1e-4, stopping_metric = "AUC", 

  ## sample 80% of rows per tree
  sample_rate = 0.8,                                                       

  ## sample 80% of columns per split
  col_sample_rate = 0.8,                                                   

  ## fix a random number generator seed for reproducibility
  seed = 1234,                                                             

  ## score every 10 trees to make early stopping reproducible (it depends on the scoring interval)
  score_tree_interval = 10)

gbm_lucky.train(x=predictors, y=response, training_frame=train, validation_frame=valid)


gbm Model Build Progress: [##################################################] 100%


This model doesn't seem to be better than the previous models:

In [11]:
perf_lucky = gbm_lucky.model_performance(valid)
print perf_lucky.auc()

0.943195266272


For this small dataset, dropping 20% of observations per tree seems too aggressive in terms of adding regularization. For larger datasets, this is usually not a bad idea. But we'll let this parameter tune freshly below, so no worries.

## Hyper-Parameter Search

Next, we'll do real hyper-parameter optimization to see if we can beat the best AUC so far (around 94%).

The key here is to start tuning some key parameters first (i.e., those that we expect to have the biggest impact on the results). From experience with gradient boosted trees across many datasets, we can state the following "rules":

1. Build as many trees (`ntrees`) as it takes until the validation set error starts increasing.
2. A lower learning rate (`learn_rate`) is generally better, but will require more trees. Using `learn_rate=0.02 `and `learn_rate_annealing=0.995` (reduction of learning rate with each additional tree) can help speed up convergence without sacrificing accuracy too much, and is great to hyper-parameter searches. For faster scans, use values of 0.05 and 0.99 instead.
3. The optimum maximum allowed depth for the trees (`max_depth`) is data dependent, deeper trees take longer to train, especially at depths greater than 10.
4. Row and column sampling (`sample_rate` and `col_sample_rate`) can improve generalization and lead to lower validation and test set errors. Good general values for large datasets are around 0.7 to 0.8 (sampling 70-80 percent of the data) for both parameters. Column sampling per tree (`col_sample_rate_per_tree`) can also be tuned. Note that it is multiplicative with `col_sample_rate`, so setting both parameters to 0.8 results in 64% of columns being considered at any given node to split.
5. For highly imbalanced classification datasets (e.g., fewer buyers than non-buyers), stratified row sampling based on response class membership can help improve predictive accuracy.  It is configured with `sample_rate_per_class` (array of ratios, one per response class in lexicographic order).
6. Most other options only have a small impact on the model performance, but are worth tuning with a Random hyper-parameter search nonetheless, if highest performance is critical.

First we want to know what value of `max_depth` to use because it has a big impact on the model training time and optimal values depend strongly on the dataset.
We'll do a quick Cartesian grid search to get a rough idea of good candidate `max_depth` values. Each model in the grid search will use early stopping to tune the number of trees using the validation set AUC, as before.
We'll use learning rate annealing to speed up convergence without sacrificing too much accuracy.

In [12]:
## Depth 10 is usually plenty of depth for most datasets, but you never know
hyper_params = {'max_depth' : range(1,30,2)}
#hyper_params = {max_depth = [4,6,8,12,16,20]} ##faster for larger datasets

#Build initial GBM Model
gbm_grid = h2o.H2OGradientBoostingEstimator(
        ## more trees is better if the learning rate is small enough 
        ## here, use "more than enough" trees - we have early stopping
        ntrees=10000,
        ## smaller learning rate is better
        ## since we have learning_rate_annealing, we can afford to start with a 
        #bigger learning rate
        learn_rate=0.05,
        ## learning rate annealing: learning_rate shrinks by 1% after every tree 
        ## (use 1.00 to disable, but then lower the learning_rate)
        learn_rate_annealing = 0.99,
        ## sample 80% of rows per tree
        sample_rate = 0.8,
        ## sample 80% of columns per split
        col_sample_rate = 0.8,
        ## fix a random number generator seed for reproducibility
        seed = 1234,
        ## score every 10 trees to make early stopping reproducible 
        #(it depends on the scoring interval)
        score_tree_interval = 10, 
        ## early stopping once the validation AUC doesn't improve by at least 0.01% for 
        #5 consecutive scoring events
        stopping_rounds = 5,
        stopping_metric = "AUC",
        stopping_tolerance = 1e-4)

#Build grid search with previously made GBM and hyper parameters
grid = h2o.H2OGridSearch(gbm_grid,hyper_params,
                         grid_id = 'depth_grid',
                         search_criteria = {'strategy': "Cartesian"})


#Train grid search
grid.train(x=predictors, 
           y=response,
           training_frame = train,
           validation_frame = valid)


gbm Grid Build Progress: [##################################################] 100%


In [13]:
## by default, display the grid search results sorted by increasing logloss (since this is a classification task)
print grid

     max_depth            model_ids              logloss
0           17   depth_grid_model_8  0.20544094075930078
1           19   depth_grid_model_9  0.20584402503242194
2           27  depth_grid_model_13  0.20627418156921704
3           11   depth_grid_model_5   0.2069364255413584
4           13   depth_grid_model_6   0.2078569528636169
5           25  depth_grid_model_12  0.20834760530631993
6            9   depth_grid_model_4  0.20842232867415922
7           29  depth_grid_model_14  0.20904163538087436
8           15   depth_grid_model_7  0.20991531457742935
9           23  depth_grid_model_11   0.2104361858121492
10          21  depth_grid_model_10  0.21069590143686837
11           7   depth_grid_model_3  0.21127939637392396
12           5   depth_grid_model_2  0.21509420086032935
13           3   depth_grid_model_1  0.21854010261642962
14           1   depth_grid_model_0  0.23892331983893703



In [14]:
## sort the grid models by decreasing AUC
sorted_grid = grid.get_grid(sort_by='auc',decreasing=True)
print(sorted_grid)

     max_depth            model_ids                 auc
0           13   depth_grid_model_6  0.9552831783601015
1           27  depth_grid_model_13  0.9547196393350239
2           17   depth_grid_model_8  0.9543251620174698
3           11   depth_grid_model_5  0.9538743307974078
4            9   depth_grid_model_4  0.9534798534798535
5           19   depth_grid_model_9  0.9534234995773457
6           25  depth_grid_model_12  0.9529726683572838
7           29  depth_grid_model_14  0.9528036066497605
8           21  depth_grid_model_10  0.9526908988447449
9           15   depth_grid_model_7  0.9526345449422373
10           7   depth_grid_model_3   0.951789236404621
11          23  depth_grid_model_11  0.9505494505494505
12           3   depth_grid_model_1   0.949084249084249
13           5   depth_grid_model_2  0.9484361792054099
14           1   depth_grid_model_0  0.9478162862778248



It appears that `max_depth` values of 9 to 27 are best suited for this dataset, which is unusally deep!

In [31]:
max_depths = sorted_grid.sorted_metric_table()['max_depth'][0:5]
new_max = int(max(max_depths, key=int))
new_min = int(min(max_depths, key=int))

print "MaxDepth", new_max
print "MinDepth", new_min

MaxDepth 27
MinDepth 9


Now that we know a good range for max_depth, we can tune all other parameters in more detail. Since we don't know what combinations of hyper-parameters will result in the best model, we'll use random hyper-parameter search to "let the machine get luckier than a best guess of any human".

In [32]:
# create hyperameter and search criteria lists (ranges are inclusive..exclusive))
hyper_params_tune = {'max_depth' : list(range(new_min,new_max+1,1)),
                'sample_rate': [x/100. for x in range(20,101)],
                'col_sample_rate' : [x/100. for x in range(20,101)],
                'col_sample_rate_per_tree': [x/100. for x in range(20,101)],
                'col_sample_rate_change_per_level': [x/100. for x in range(90,111)],
                'min_rows': [2**x for x in range(0,int(math.log(train.nrow,2)-1)+1)],
                'nbins': [2**x for x in range(4,11)],
                'nbins_cats': [2**x for x in range(4,13)],
                'min_split_improvement': [0,1e-8,1e-6,1e-4],
                'histogram_type': ["UniformAdaptive","QuantilesGlobal","RoundRobin"]}
search_criteria_tune = {'strategy': "RandomDiscrete",
                   'max_runtime_secs': 3600,  ## limit the runtime to 60 minutes
                   'max_models': 100,  ## build no more than 100 models
                   'seed' : 1234,
                   'stopping_rounds' : 5,
                   'stopping_metric' : "AUC",
                   'stopping_tolerance': 1e-3
                   }

In [33]:
gbm_final_grid = h2o.H2OGradientBoostingEstimator(distribution='bernoulli',
                    ## more trees is better if the learning rate is small enough 
                    ## here, use "more than enough" trees - we have early stopping
                    ntrees=10000,
                    ## smaller learning rate is better
                    ## since we have learning_rate_annealing, we can afford to start with a 
                    #bigger learning rate
                    learn_rate=0.05,
                    ## learning rate annealing: learning_rate shrinks by 1% after every tree 
                    ## (use 1.00 to disable, but then lower the learning_rate)
                    learn_rate_annealing = 0.99,
                    ## score every 10 trees to make early stopping reproducible 
                    #(it depends on the scoring interval)
                    score_tree_interval = 10,
                    ## fix a random number generator seed for reproducibility
                    seed = 1234,
                    ## early stopping once the validation AUC doesn't improve by at least 0.01% for 
                    #5 consecutive scoring events
                    stopping_rounds = 5,
                    stopping_metric = "AUC",
                    stopping_tolerance = 1e-4)
            
#Build grid search with previously made GBM and hyper parameters
final_grid = h2o.H2OGridSearch(gbm_final_grid, hyper_params = hyper_params_tune,
                                    grid_id = 'final_grid',
                                    search_criteria = search_criteria_tune)
#Train grid search
final_grid.train(x=predictors, 
           y=response,
           ## early stopping based on timeout (no model should take more than 1 hour - modify as needed)
           max_runtime_secs = 3600, 
           training_frame = train,
           validation_frame = valid)

print final_grid


gbm Grid Build Progress: [##################################################] 100%
      col_sample_rate col_sample_rate_change_per_level  \
0                0.49                             1.04   
1                0.35                             1.09   
2                0.73                              0.9   
3                0.97                             0.96   
4                 0.5                             1.02   
5                 0.6                              1.0   
6                 0.5                             0.94   
7                0.81                             0.94   
8                0.63                              1.0   
9                 0.4                             1.01   
10               0.65                             1.08   
11                0.2                             0.96   
12               0.61                             1.04   
13               0.45                             1.03   
14               0.85                         

We can see that the best models have even better validation AUCs than our previous best models, so the random grid search was successful!

In [34]:
## Sort the grid models by AUC
sorted_final_grid = final_grid.get_grid(sort_by='auc',decreasing=True)

print sorted_final_grid

      col_sample_rate col_sample_rate_change_per_level  \
0                0.73                              0.9   
1                0.49                             1.04   
2                0.92                             0.93   
3                 0.5                             1.02   
4                0.35                             1.09   
5                 0.7                             1.02   
6                0.81                             0.94   
7                0.61                             1.04   
8                 0.4                             1.01   
9                0.97                             0.96   
10               0.91                             0.96   
11                0.5                             0.94   
12               0.51                              1.1   
13               0.66                             0.96   
14               0.96                              1.1   
15               0.71                              0.9   
16            

You can also see the results of the grid search in [Flow](http://localhost:54321/):
![alt text](./final_grid.png "Final Grid Search Results in Flow")

## Model Inspection and Final Test Set Scoring

Let's see how well the best model of the grid search (as judged by validation set AUC) does on the held out test set:

In [37]:
#Get the best model from the list (the model name listed at the top of the table)
best_model = h2o.get_model(sorted_final_grid.sorted_metric_table()['model_ids'][0])
performance_best_model = best_model.model_performance(test)
print performance_best_model.auc()

0.97878948064


Good news. It does as well on the test set as on the validation set, so it looks like our best GBM model generalizes well to the unseen test set:

We can inspect the winning model's parameters:

In [38]:
params_list = []
for key, value in best_model.params.iteritems():
    params_list.append(str(key)+" = "+str(value['actual']))
params_list

['learn_rate = 0.05',
 'fold_column = None',
 'col_sample_rate_per_tree = 0.6',
 'learn_rate_annealing = 0.99',
 'score_tree_interval = 10',
 'sample_rate_per_class = None',
 'seed = 1234',
 'keep_cross_validation_predictions = False',
 "model_id = {u'URL': u'/3/Models/final_grid_model_45', u'_exclude_fields': u'', u'type': u'Key<Model>', u'name': u'final_grid_model_45', u'__meta': {u'schema_name': u'ModelKeyV3', u'schema_version': 3, u'schema_type': u'Key<Model>'}}",
 'nfolds = 0',
 'max_abs_leafnode_pred = 1.79769313486e+308',
 'offset_column = None',
 'quantile_alpha = 0.5',
 'stopping_tolerance = 0.0001',
 'fold_assignment = AUTO',
 "training_frame = {u'URL': u'/3/Frames/train.hex', u'_exclude_fields': u'', u'type': u'Key<Frame>', u'name': u'train.hex', u'__meta': {u'schema_name': u'FrameKeyV3', u'schema_version': 3, u'schema_type': u'Key<Frame>'}}",
 'max_runtime_secs = 3512.106',
 'checkpoint = None',
 'balance_classes = False',
 'r2_stopping = 0.999999',
 "validation_frame = {u'

Now we can confirm that these parameters are generally sound, by building a GBM model on the whole dataset (instead of the 60%) and using internal 5-fold cross-validation (re-using all other parameters including the seed):

In [39]:
gbm = h2o.get_model(sorted_final_grid.sorted_metric_table()['model_ids'][0])
#get the parameters from the Random grid search model and modify them slightly
params = gbm.params
new_params = {"nfolds":5, "model_id":None}
for key in new_params.keys():
    params[key]['actual'] = new_params[key] 
gbm_best = h2o.H2OGradientBoostingEstimator()
for key in params.keys():
    if key in dir(gbm_best) and getattr(gbm_best,key) != params[key]['actual']:
        setattr(gbm_best,key,params[key]['actual']) 

In [40]:
gbm_best.train(x=predictors, y=response, training_frame=df)


gbm Model Build Progress: [##################################################] 100%


In [41]:
print gbm_best.cross_validation_metrics_summary()


Cross-Validation Metrics Summary: 


mean       sd          cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
-----------------------  ---------  ----------  ------------  ------------  ------------  ------------  ------------
accuracy                 0.9444     0.00643399  0.940075      0.933579      0.937984      0.956693      0.953668
auc                      0.970471   0.00700594  0.963998      0.958452      0.965561      0.983979      0.980366
err                      0.0556001  0.00643399  0.0599251     0.0664207     0.0620155     0.0433071     0.046332
err_count                14.6       1.87617     16            18            16            11            12
f0point5                 0.948325   0.0102195   0.94697       0.921053      0.956221      0.961969      0.955414
f1                       0.923726   0.0100577   0.925926      0.903226      0.912088      0.939891      0.9375
f2                       0.900554   0.0133487   0.905797      0.886076      0.871849      0.918803      0.920245
lift_top_group           2.62587    0.0998947   2.38393       2.82292       2.63265       2.67368       2.61616
logloss                  0.18693    0.0142185   0.203403      0.2099        0.193357      0.15595       0.172042
max_per_class_error      0.11417    0.016118    0.107143      0.125         0.153061      0.0947368     0.0909091
mcc                      0.88231    0.0141451   0.877423      0.853773      0.870751      0.907743      0.901859
mean_per_class_accuracy  0.933148   0.00803933  0.933525      0.920357      0.920344      0.946342      0.94517
mean_per_class_error     0.0668521  0.00803933  0.0664747     0.0796429     0.0796556     0.0536577     0.0548295
mse                      0.0501457  0.00427637  0.0544056     0.0572033     0.0529171     0.0411026     0.0451
precision                0.965596   0.0130557   0.961538      0.933333      0.988095      0.977273      0.967742
r2                       0.787065   0.018719    0.776583      0.749936      0.775359      0.824444      0.809005
recall                   0.88583    0.016118    0.892857      0.875         0.846939      0.905263      0.909091
specificity              0.980466   0.0069456   0.974194      0.965714      0.99375       0.987421      0.98125

It looks like the winning model performs slightly better on the validation and test sets than during cross-validation on the training set as the mean AUC on the 5 folds is estimated to be only 97.04%, but with a fairly large standard deviation of 0.7%. For small datasets, such a large variance is not unusual. To get a better estimate of model performance, the Random hyper-parameter search could have used `nfolds = 5` (or 10, or similar) in combination with 80% of the data for training (i.e., not holding out a validation set, but only the final test set). However, this would take more time, as `nfolds+1` models will be built for every set of parameters.

Instead, to save time, let's just scan through the top 5 models and cross-validate their parameters with `nfolds=5` on the entire dataset:

In [42]:
for i in range(5): 
    gbm = h2o.get_model(sorted_final_grid.sorted_metric_table()['model_ids'][i])
    #get the parameters from the Random grid search model and modify them slightly
    params = gbm.params
    new_params = {"nfolds":5, "model_id":None}
    for key in new_params.keys():
        params[key]['actual'] = new_params[key]
    new_model = h2o.H2OGradientBoostingEstimator()
    for key in params.keys():
        if key in dir(new_model) and getattr(new_model,key) != params[key]['actual']:
            setattr(new_model,key,params[key]['actual'])
    new_model.train(x = predictors, y = response, training_frame = df)  
    cv_summary = new_model.cross_validation_metrics_summary().as_data_frame()
    print(gbm.model_id)
    print(cv_summary.iloc[1]) ## AUC


gbm Model Build Progress: [##################################################] 100%
final_grid_model_45
                       auc
mean            0.97047126
sd            0.0070059407
cv_1_valid       0.9639977
cv_2_valid       0.9584524
cv_3_valid       0.9655612
cv_4_valid       0.9839788
cv_5_valid       0.9803662
Name: 1, dtype: object

gbm Model Build Progress: [##################################################] 100%
final_grid_model_68
                      auc
mean            0.9719337
sd            0.006935956
cv_1_valid      0.9698157
cv_2_valid     0.95497024
cv_3_valid      0.9716199
cv_4_valid     0.98106587
cv_5_valid       0.982197
Name: 1, dtype: object

gbm Model Build Progress: [##################################################] 100%
final_grid_model_96
                      auc
mean            0.9722442
sd            0.005797062
cv_1_valid      0.9639401
cv_2_valid     0.96068454
cv_3_valid       0.978125
cv_4_valid       0.978484
cv_5_valid      0.9799874
Name: 1

The avid reader might have noticed that we just implicitly did further parameter tuning using the "final" test set (which is part of the entire dataset `df`), which is not good practice - one is not supposed to use the "final" test set more than once. Hence, we're not going to pick a different "best" model, but we're just learning about the variance in AUCs. It turns out, for this tiny dataset, that the variance is rather large, which is not surprising.

Keeping the same "best" model, we can make test set predictions as follows:

In [43]:
preds = best_model.predict(test)
preds.head()


gbm prediction Progress: [##################################################] 100%


predict p0 p1 
 0 0.948394 0.0516064 
 0 0.940819 0.0591814 
 0 0.924897 0.0751026 
 1 0.0156249 0.984375 
 1 0.0116196 0.98838 
 0 0.854808 0.145192 
 1 0.0440646 0.955935 
 1 0.0115973 0.988403 
 1 0.0544365 0.945563 
 0 0.919649 0.0803507

Note that the label (survived or not) is predicted as well (in the first predict column), and it uses the threshold with the highest F1 score (here: 0.528098) to make labels from the probabilities for survival (`p1`). The probability for death (`p0`) is given for convenience, as it is just `1-p1`.

In [44]:
best_model.model_performance(valid)


ModelMetricsBinomial: gbm
** Reported on test data. **

MSE: 0.0490254194251
R^2: 0.792582001197
LogLoss: 0.183618009053
Mean Per-Class Error: 0.0666666666667
AUC: 0.97235841082
Gini: 0.94471682164

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.492182327674: 


0    1    Error    Rate
-----  ---  ---  -------  ------------
0      169  0    0        (0.0/169.0)
1      14   91   0.1333   (14.0/105.0)
Total  183  91   0.0511   (14.0/274.0)


Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.492182     0.928571  90
max f2                       0.161259     0.907407  119
max f0point5                 0.492182     0.970149  90
max accuracy                 0.492182     0.948905  90
max precision                0.988563     1         0
max recall                   0.0501451    1         199
max specificity              0.988563     1         0
max absolute_MCC             0.492182     0.894631  90
max min_per_class_accuracy   0.215866     0.905325  111
max mean_per_class_accuracy  0.492182     0.933333  90


Gains/Lift Table: Avg response rate: 38.32 %



group    cumulative_data_fraction    lower_threshold    lift      cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  --------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0109489                   0.987885           2.60952   2.60952            1                1                           0.0285714       0.0285714                  160.952   160.952
    2        0.0218978                   0.987106           2.60952   2.60952            1                1                           0.0285714       0.0571429                  160.952   160.952
    3        0.0328467                   0.986676           2.60952   2.60952            1                1                           0.0285714       0.0857143                  160.952   160.952
    4        0.040146                    0.9859             2.60952   2.60952            1                1                           0.0190476       0.104762                   160.952   160.952
    5        0.0510949                   0.985696           2.60952   2.60952            1                1                           0.0285714       0.133333                   160.952   160.952
    6        0.10219                     0.979859           2.60952   2.60952            1                1                           0.133333        0.266667                   160.952   160.952
    7        0.149635                    0.973184           2.60952   2.60952            1                1                           0.12381         0.390476                   160.952   160.952
    8        0.20073                     0.954807           2.60952   2.60952            1                1                           0.133333        0.52381                    160.952   160.952
    9        0.29927                     0.892458           2.60952   2.60952            1                1                           0.257143        0.780952                   160.952   160.952
    10       0.40146                     0.232713           1.21156   2.25368            0.464286         0.863636                    0.12381         0.904762                   21.1565   125.368
    11       0.5                         0.116312           0.483245  1.90476            0.185185         0.729927                    0.047619        0.952381                   -51.6755  90.4762
    12       0.59854                     0.0723806          0.193298  1.623              0.0740741        0.621951                    0.0190476       0.971429                   -80.6702  62.2997
    13       0.70073                     0.0526387          0.186395  1.41349            0.0714286        0.541667                    0.0190476       0.990476                   -81.3605  41.3492
    14       0.79927                     0.0440208          0.096649  1.25114            0.037037         0.479452                    0.00952381      1                          -90.3351  25.1142
    15       0.89781                     0.036913           0         1.11382            0                0.426829                    0               1                          -100      11.3821
    16       1                           0.0165575          0         1                  0                0.383212                    0               1                          -100      0

You can also see the "best" model in more detail in [Flow](http://localhost:54321/):
![alt text](./best_gbm1.png "Best GBM from Grid Search in Flow")
![alt text](./best_gbm2.png "Best GBM from Grid Search in Flow")

The model and the predictions can be saved to file as follows:

In [45]:
h2o.save_model(best_model, "/tmp/bestModel.csv", force=True)
h2o.export_file(preds, "/tmp/bestPreds.csv", force=True)


Export File Progress: [##################################################] 100%


In [32]:
#h2o.download_pojo(best_model)

The model can also be exported as a plain old Java object (POJO) for H2O-independent (standalone/Storm/Kafka/UDF) scoring in any Java environment.

```
/*
 Licensed under the Apache License, Version 2.0
    http://www.apache.org/licenses/LICENSE-2.0.html

  AUTOGENERATED BY H2O at 2016-07-17T18:38:50.337-07:00
  3.8.3.3

  Standalone prediction code with sample test data for GBMModel named final_grid_model_45

  How to download, compile and execute:
      mkdir tmpdir
      cd tmpdir
      curl http://127.0.0.1:54321/3/h2o-genmodel.jar > h2o-genmodel.jar
      curl http://127.0.0.1:54321/3/Models.java/final_grid_model_45 > final_grid_model_45.java
      javac -cp h2o-genmodel.jar -J-Xmx2g -J-XX:MaxPermSize=128m final_grid_model_45.java

     (Note:  Try java argument -XX:+PrintCompilation to show runtime JIT compiler behavior.)
*/
import java.util.Map;
import hex.genmodel.GenModel;
import hex.genmodel.annotations.ModelPojo;

...
class final_grid_model_45_Tree_0_class_0 {
  static final double score0(double[] data) {
    double pred =      (Double.isNaN(data[1]) || !GenModel.bitSetContains(GRPSPLIT0, 0, data[1 /* sex */]) ? 
         (Double.isNaN(data[7]) || !GenModel.bitSetContains(GRPSPLIT1, 13, data[7 /* cabin */]) ? 
             (Double.isNaN(data[7]) || !GenModel.bitSetContains(GRPSPLIT2, 9, data[7 /* cabin */]) ? 
                 (Double.isNaN(data[7]) || !GenModel.bitSetContains(GRPSPLIT3, 9, data[7 /* cabin */]) ? 
                     (data[2 /* age */] <1.4174492f ? 
                        0.13087687f : 
                         (Double.isNaN(data[7]) || !GenModel.bitSetContains(GRPSPLIT4, 9, data[7 /* cabin */]) ? 
                             (Double.isNaN(data[3]) || data[3 /* sibsp */] <1.000313f ? 
                                 (data[6 /* fare */] <7.91251f ? 
                                     (Double.isNaN(data[5]) || data[5 /* ticket */] <368744.5f ? 
                                        -0.08224204f : 
                                         (Double.isNaN(data[2]) || data[2 /* age */] <13.0f ? 
                                            -0.028962314f : 
                                            -0.08224204f)) : 
                                     (Double.isNaN(data[7]) || !GenModel.bitSetContains(GRPSPLIT5, 9, data[7 /* cabin */]) ? 
                                         (data[6 /* fare */] <7.989957f ? 
                                             (Double.isNaN(data[3]) || data[3 /* sibsp */] <0.0017434144f ? 
                                                0.07759714f : 
                                                0.13087687f) : 
                                             (data[6 /* fare */] <12.546303f ? 
                                                -0.07371729f : 
                                                 (Double.isNaN(data[4]) || data[4 /* parch */] <1.0020853f ? 
                                                    -0.037374903f : 
                                                    -0.08224204f))) : 
                                        0.0f)) : 
                                -0.08224204f) : 
                            0.0f)) : 
                    0.0f) : 
                -0.08224204f) : 
            -0.08224204f) :  
...
```

## Ensembling Techniques

After learning above that the variance of the test set AUC of the top few models was rather large, we might be able to turn this into our advantage by using ensembling techniques. The simplest one is taking the average of the predictions (survival probabilities) of the top `k` grid search model predictions (here, we use `k=10`):

In [46]:
prob = None
k=10
for i in range(0,k): 
    gbm = h2o.get_model(sorted_final_grid.sorted_metric_table()['model_ids'][i])
    if (prob is None):
        prob = gbm.predict(test)["p1"]
    else:
        prob = prob + gbm.predict(test)["p1"]
prob = prob/k


gbm prediction Progress: [##################################################] 100%

gbm prediction Progress: [##################################################] 100%

gbm prediction Progress: [##################################################] 100%

gbm prediction Progress: [##################################################] 100%

gbm prediction Progress: [##################################################] 100%

gbm prediction Progress: [##################################################] 100%

gbm prediction Progress: [##################################################] 100%

gbm prediction Progress: [##################################################] 100%

gbm prediction Progress: [##################################################] 100%

gbm prediction Progress: [##################################################] 100%


We now have a blended probability of survival for each person on the Titanic.

In [47]:
prob.head()

p1 
 0.0596246 
 0.0511568 
 0.115645 
 0.977491 
 0.981366 
 0.210353 
 0.948326 
 0.976989 
 0.944669 
 0.107771

We can bring those ensemble predictions to our Python session's memory space and use other Python packages.

In [48]:
from sklearn.metrics import roc_auc_score
# convert prob and test[response] h2oframes to pandas' frames and then convert them each to numpy array
np_array_prob = prob.as_data_frame().as_matrix()
np_array_test = test[response].as_data_frame().as_matrix()
probInPy = np_array_prob
labeInPy = np_array_test
# compare true scores (test[response]) to probability scores (prob)
roc_auc_score(labeInPy, probInPy)

0.98202722347033167

This simple blended ensemble test set prediction has an even higher AUC than the best single model, but we need to do more validation studies, ideally using cross-validation. We leave this as an exercise for the reader - take the parameters of the top `10` models, retrain them with `nfolds=5` on the full dataset, set `keep_holdout_predictions=True` and sum up their predicted probabilities, then score that with sklearn's roc_auc_score as shown above.

For more sophisticated ensembling approaches, such as stacking via a superlearner, we refer to the [H2O Ensemble](https://github.com/h2oai/h2o-3/tree/master/h2o-r/ensemble) github page.

## Summary
We learned how to build H2O GBM models for a binary classification task on a small but realistic dataset with numerical and categorical variables, with the goal to maximize the AUC (ranges from 0.5 to 1). We first established a baseline with the default model, then carefully tuned the remaining hyper-parameters without "too much" human guess-work. We used both Cartesian and Random hyper-parameter searches to find good models. We were able to get the AUC on a holdout test set from the low 94% range with the default model to the mid 97% range after tuning, and to above 98% with some simple ensembling technique known as blending. We performed simple cross-validation variance analysis to learn that results were slightly "lucky" due to the specific train/valid/test set splits, and settled to expect 97% AUCs instead.

Note that this script and the findings therein are directly transferrable to large datasets on distributed clusters including Spark/Hadoop environments.

More information can be found here [http://www.h2o.ai/docs/](http://www.h2o.ai/docs/).